util/http/retry.go

```diff
@@ -89,3 +92,39 @@ func RetryStrategy(ctx context.Context, resp *http.Response, err error) (bool, e
 	}
 	return retry, err2
 }
+
+// ShouldReauthenticate takes an HTTP response and determines whether or not
+// it warrants re-authentication.
+func ShouldReauthenticate(resp *http.Response) bool {
+	defer resp.Body.Close()
+
+	switch resp.StatusCode {
+	case http.StatusUnauthorized:
+		return true
+	case http.StatusForbidden:
+		// Although in most cases 403 responses represent authorization issues that generally
+		// cannot be resolved by re-authentication, some registries like ECR, will return a 403 on
+		// credential expiration. (ref https://docs.aws.amazon.com/AmazonECR/latest/userguide/common-errors-docker.html#error-403)
+		// In the case of ECR, the response body is structured according to the error format defined in the
+		// Docker v2 API spec. (ref https://distribution.github.io/distribution/spec/api/#errors).
+
+		// We will attempt to parse the response body as a `docker.Errors`. If it can be parsed
+		// we will ensure that the `Message` represents token expiration.
+
+		var errs docker.Errors
+		respBody := bytes.NewBuffer([]byte{})
+		_, err := io.Copy(respBody, resp.Body)
```

Comment:
Not sure if we should employ some retry policy here?

Reply:
Maybe an easier way is to do this?
```
json.NewDecoder(resp.Body).Decode(&errs)
```